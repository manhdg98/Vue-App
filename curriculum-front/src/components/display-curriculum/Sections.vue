<template>
  <v-row>
    <v-col>
      <v-expansion-panels multiple>
        <v-expansion-panel
          v-for="(section, i) in selectedCurriculum.sections"
          :key="i"
        >
          <v-expansion-panel-header>
            <div class="d-flex align-center justify-space-between">
              <span class="section-header-text">
                Section {{ i + 1 }} - {{ section.name }}
              </span>
              <span>
                <v-icon
                  v-if="canEdit()"
                  color="gray lighten-1"
                  @click.stop="openEditSectionDialog(i)"
                >
                  mdi-pencil-box-outline
                </v-icon>
                <v-icon
                  v-if="canEdit()"
                  color="error lighten-1"
                  @click.stop="toggleConfirmDelete(i)"
                >
                  mdi-close-box-outline
                </v-icon>
              </span>
            </div>
          </v-expansion-panel-header>
          <v-expansion-panel-content>
            <v-list
              subheader
              two-line
              flat
            >
              <v-subheader>
                Resources
                <v-icon
                  v-if="canEdit()"
                  color="success lighten-1"
                  class="ml-1"
                  @click="toggleDialog('resource', i)"
                >
                  mdi-plus-box-outline
                </v-icon>
              </v-subheader>

              <v-list-item-group
                multiple
              >
                <v-list-item
                  v-for="(resource, j) in section.resources"
                  :key="resource + j"
                >
                  <v-checkbox
                    v-if="canEdit()"
                    color="primary"
                    v-model="resource.isCompleted"
                    @change="toggleComplete('resources', i, j)"
                  />

                  <v-list-item-content>
                    <v-list-item-title class="hover-icon-container">
                      <router-link to="/">
                        {{ resource.name }}
                      </router-link>
                      <v-icon
                        v-if="canEdit()"
                        color="gray lighten-1"
                        class="togglable-icon"
                        @click="toggleDialog('resource', i, j)"
                      >
                        mdi-pencil-box-outline
                      </v-icon>
                      <v-icon
                        v-if="canEdit()"
                        color="error lighten-1"
                        class="ml-1 togglable-icon"
                        @click="removeItem('resources', i, j)"
                      >
                        mdi-close-box-outline
                      </v-icon>
                    </v-list-item-title>
                  </v-list-item-content>
                </v-list-item>
              </v-list-item-group>
            </v-list>
            <v-list
              subheader
              two-line
              flat
            >
              <v-subheader>
                Projects
                <v-icon
                  v-if="canEdit()"
                  color="success lighten-1"
                  class="ml-1"
                  @click="toggleDialog('project', i)"
                >
                  mdi-plus-box-outline
                </v-icon>
              </v-subheader>

              <v-list-item-group
                multiple
              >
                <v-list-item
                  v-for="(project, k) in section.projects"
                  :key="project + k"
                >
                  <v-checkbox
                    color="primary"
                    v-model="project.isCompleted"
                    @change="toggleComplete('projects', i, k)"
                  />

                  <v-list-item-content>
                    <v-list-item-title class="hover-icon-container">
                      <router-link to="/">
                        {{ project.name }}
                      </router-link>
                      <v-icon
                        v-if="canEdit()"
                        color="gray lighten-1"
                        class="togglable-icon"
                        @click="toggleDialog('project', i, k)"
                      >
                        mdi-pencil-box-outline
                      </v-icon>
                      <v-icon
                        v-if="canEdit()"
                        color="error lighten-1"
                        class="ml-1 togglable-icon"
                        @click.stop="removeItem('projects', i, j)"
                      >
                        mdi-close-box-outline
                      </v-icon>
                    </v-list-item-title>
                  </v-list-item-content>
                </v-list-item>
              </v-list-item-group>
            </v-list>
          </v-expansion-panel-content>
        </v-expansion-panel>
      </v-expansion-panels>
    </v-col>
    <transition name="slide-fade">
      <AddItemDialog
        :dialog="dialog"
        :toggleDialog="toggleDialog"
        :saveItem="saveItem"
      />
    </transition>
    <v-dialog v-model="showDeleteDialog" persistent max-width="290">
      <v-card>
        <v-card-title class="headline">Do you really want to delete this section?</v-card-title>
        <v-card-actions>
          <v-spacer></v-spacer>
          <v-btn color="green darken-1" text @click="showDeleteDialog = false">Cancel</v-btn>
          <v-btn color="green darken-1" text @click="confirmDelete">Confirm</v-btn>
        </v-card-actions>
      </v-card>
    </v-dialog>
    <v-dialog v-model="showEditSectionDialog" persistent max-width="400">
      <v-card>
        <v-card-title class="headline">Edit Section Information</v-card-title>
        <v-card-text>
          <v-row no-gutters>
            <v-col cols="12">
              <v-text-field
                placeholder="Section Name"
                v-model="sectionEditing.name"
              />
            </v-col>
          </v-row>
          <v-row no-gutters>
            <v-col cols="12">
              <v-text-field
                placeholder="Section Goal"
                v-model="sectionEditing.goal"
              />
            </v-col>
          </v-row>
        </v-card-text>
        <v-card-actions>
          <v-spacer></v-spacer>
          <v-btn color="green darken-1" text @click="showEditSectionDialog = false">Cancel</v-btn>
          <v-btn color="green darken-1" text @click="updateSection">Save</v-btn>
        </v-card-actions>
      </v-card>
    </v-dialog>
  </v-row>
</template>

<script>
import AddItemDialog from './AddItemDialog.vue'
import { mapActions } from 'vuex'

export default {
  props: {
    dialog: Object,
    selectedCurriculum: Object,
    toggleComplete: Function,
    toggleDialog: Function,
    saveItem: Function,
    editItem: Function,
    removeItem: Function,
    canEdit: Function
  },
  data () {
    return {
      showDeleteDialog: false,
      showEditSectionDialog: false,
      selectedIndex: null,
      sectionEditing: {}
    }
  },
  components: {
    AddItemDialog
  },
  methods: {
    ...mapActions(['deleteSection', 'patchSection']),
    toggleConfirmDelete (index) {
      this.selectedIndex = index
      this.showDeleteDialog = !this.showDeleteDialog
    },
    openEditSectionDialog (index) {
      this.selectedIndex = index
      this.sectionEditing = this.selectedCurriculum.sections[index]
      this.showEditSectionDialog = !this.showEditSectionDialog
    },
    closeEditSectionDialog (index) {
      this.sectionEditing = {}
      this.showEditSectionDialog = false
    },
    confirmDelete () {
      const section = this.selectedCurriculum.sections[this.selectedIndex]
      const payload = {
        curriculumId: this.selectedCurriculum._id,
        sectionId: section._id
      }
      this.deleteSection(payload)
      this.toggleConfirmDelete()
    },
    updateSection () {
      const payload = {
        curriculumId: this.selectedCurriculum._id,
        section: this.sectionEditing
      }
      this.patchSection(payload)
      this.closeEditSectionDialog()
    }
  }
}
</script>
