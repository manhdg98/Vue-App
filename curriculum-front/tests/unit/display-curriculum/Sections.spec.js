import { mount } from '@vue/test-utils'
import Sections from '@/components/display-curriculum/Sections.vue'

const transitionStub = () => ({
  render: function (h) {
    return this.$options._renderChildren
  }
})

describe('Sections.vue', () => {
  test('if snapshot is correct', () => {
    const wrapper = mount(Sections, {
      stubs: {
        transition: transitionStub(),
        AddItemDialog: {
          template: '<div></div>'
        },
        ThemeProvider: {
          template: '<div></div>'
        },
        VDialog: {
          template: '<div></div>'
        }
      },
      propsData: {
        dialog: {
          type: '',
          show: false,
          name: '',
          url: '',
          isEditing: false,
          itemIndex: null,
          sectionIndex: null
        },
        selectedCurriculum: {
          sections: [
            {
              name: 'section 1',
              projects: [
                {
                  name: 'Project 1'
                }
              ],
              resources: [
                {
                  name: 'Resource 1'
                }
              ]
            },
            {
              name: 'section 2',
              projects: [
                {
                  name: 'Project 2'
                }
              ],
              resources: [
                {
                  name: 'Resource 2'
                }
              ]
            }
          ]
        },
        toggleComplete: () => {},
        toggleDialog: () => {},
        saveItem: () => {},
        editItem: () => {},
        removeItem: () => {}
      }
    })
    expect(wrapper).toMatchSnapshot()
  })
})
