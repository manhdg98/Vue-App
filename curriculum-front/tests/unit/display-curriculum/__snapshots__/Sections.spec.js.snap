// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sections.vue if snapshot is correct 1`] = `
<div class="row">
  <div class="col">
    <div class="v-item-group theme--light v-expansion-panels">
      <div aria-expanded="false" class="v-expansion-panel"><button type="button" class="v-expansion-panel-header">
          <div class="d-flex align-center justify-space-between"><span class="section-header-text">
              Section 1 - section 1
            </span> <span><i aria-hidden="true" class="v-icon notranslate mdi mdi-pencil-box-outline theme--light gray--text text--lighten-1"></i> <i role="button" class="v-icon notranslate v-icon--link mdi mdi-close-box-outline theme--light error--text text--lighten-1"></i></span></div>
          <div class="v-expansion-panel-header__icon"><i aria-hidden="true" class="v-icon notranslate material-icons theme--light">$expand</i></div>
        </button>
        <div class="v-expansion-panel-content" style="display: none;"></div>
      </div>
      <div aria-expanded="false" class="v-expansion-panel"><button type="button" class="v-expansion-panel-header">
          <div class="d-flex align-center justify-space-between"><span class="section-header-text">
              Section 2 - section 2
            </span> <span><i aria-hidden="true" class="v-icon notranslate mdi mdi-pencil-box-outline theme--light gray--text text--lighten-1"></i> <i role="button" class="v-icon notranslate v-icon--link mdi mdi-close-box-outline theme--light error--text text--lighten-1"></i></span></div>
          <div class="v-expansion-panel-header__icon"><i aria-hidden="true" class="v-icon notranslate material-icons theme--light">$expand</i></div>
        </button>
        <div class="v-expansion-panel-content" style="display: none;"></div>
      </div>
    </div>
  </div>
  <div dialog="[object Object]" toggledialog="() => {}" saveitem="() => {}" name="slide-fade"></div>
  <div persistent="" max-width="290"></div>
</div>
`;
